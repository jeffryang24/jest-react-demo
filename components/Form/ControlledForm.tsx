import { useEffect, useState } from 'react';
import { Form, FormInputProps, FormTextAreaProps } from 'semantic-ui-react';

import { getPeople } from '../../services/PeopleService';

import useTimer from '../../hooks/useTimer';

interface Props {
  age: string;
  description: string;
  fullName: string;
  onAgeChange?: FormInputProps['onChange'];
  onDescriptionChange?: FormTextAreaProps['onChange'];
  onFullNameChange?: FormInputProps['onChange'];
}

export default function ControlledForm(props: Props) {
  const {
    age,
    description,
    fullName,
    onAgeChange,
    onDescriptionChange,
    onFullNameChange,
  } = props;

  const [people, setPeople] = useState([]);
  const [timer, timerUtil] = useTimer({ startFrom: 5 });

  useEffect(() => {
    console.log('Hello from controlled form!');

    async function fetchPeople() {
      const result = await getPeople();
      setPeople(result);
    }

    fetchPeople();

    return () => {
      console.log('Bye-bye from controlled form!');
    };
  }, []);

  return (
    <Form>
      <Form.Input
        id={'fullName'}
        label={'Full Name'}
        placeholder={'Insert Full Name'}
        value={fullName}
        onChange={onFullNameChange}
      />

      <Form.Input
        id={'age'}
        label={'Age'}
        placeholder={'Insert Age'}
        type={'number'}
        min={1}
        value={age}
        onChange={onAgeChange}
      />

      <Form.TextArea
        id={'description'}
        label={'Description'}
        placeholder={'Insert Description'}
        style={{ minHeight: 150 }}
        value={description}
        onChange={onDescriptionChange}
      />

      <p>
        {'Timer ends in: '}
        <span>{timer.currentDuration}</span>
        <span style={{ marginLeft: 8 }}>
          {timer.currentDuration === 0 ? 'Ringing' : null}
        </span>
      </p>

      <Form.Group>
        <Form.Button
          type={'button'}
          color={'green'}
          onClick={timerUtil.startTimer}
        >
          {'Start Timer'}
        </Form.Button>

        <Form.Button
          type={'button'}
          color={'yellow'}
          onClick={timerUtil.pauseTimer}
        >
          {'Pause Timer'}
        </Form.Button>

        <Form.Button
          type={'button'}
          color={'red'}
          onClick={timerUtil.stopTimer}
        >
          {'Stop Timer'}
        </Form.Button>
      </Form.Group>

      <code data-testId={'peopleLength'}>{people.length}</code>
    </Form>
  );
}
