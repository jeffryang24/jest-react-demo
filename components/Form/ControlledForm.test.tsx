import { ChangeEvent, ComponentPropsWithoutRef } from 'react';
import { InputOnChangeData } from 'semantic-ui-react';
import { fireEvent, render } from '@testing-library/react';

import ControlledForm from './ControlledForm';

type ControlledFormProps = ComponentPropsWithoutRef<typeof ControlledForm>;

describe('Components - Form - Controlled Form', () => {
  function createProps(
    customProps?: Partial<ControlledFormProps>
  ): ControlledFormProps {
    return {
      age: '23',
      description: 'Lorem ipsum dolor sip amet',
      fullName: 'John Doe',
      ...customProps,
    };
  }

  function renderForm(customProps?: Partial<ControlledFormProps>) {
    const props = createProps(customProps);
    return render(<ControlledForm {...props} />);
  }

  it('should render the form gracefully - snapshot', () => {
    const { baseElement } = renderForm();

    expect(baseElement).toMatchSnapshot();
  });

  it('should show the correct value based on the props', () => {
    const defaultProps = createProps();
    const { getByDisplayValue } = renderForm();

    expect(getByDisplayValue(defaultProps.age)).toBeTruthy();
    expect(getByDisplayValue(defaultProps.description)).toBeTruthy();
    expect(getByDisplayValue(defaultProps.fullName)).toBeTruthy();
  });

  it('should show the correct changed value on onFullNameChange', () => {
    const onFullNameChange = jest.fn(
      (_e: ChangeEvent<HTMLInputElement>, data: InputOnChangeData) => data.value
    );

    const { getByLabelText } = renderForm({
      fullName: '',
      onFullNameChange,
    });

    fireEvent.change(getByLabelText('Full Name'), {
      target: {
        value: 'CJ',
      },
    });
    expect(onFullNameChange).toHaveBeenCalled();
  });
});
