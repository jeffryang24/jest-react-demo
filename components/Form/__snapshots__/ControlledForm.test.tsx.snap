// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components - Form - Controlled Form should render the form gracefully - snapshot 1`] = `
<body>
  <div>
    <form
      class="ui form"
    >
      <div
        class="field"
      >
        <label
          for="fullName"
        >
          Full Name
        </label>
        <div
          class="ui input"
        >
          <input
            id="fullName"
            placeholder="Insert Full Name"
            type="text"
            value="John Doe"
          />
        </div>
      </div>
      <div
        class="field"
      >
        <label
          for="age"
        >
          Age
        </label>
        <div
          class="ui input"
        >
          <input
            id="age"
            min="1"
            placeholder="Insert Age"
            type="number"
            value="23"
          />
        </div>
      </div>
      <div
        class="field"
      >
        <label
          for="description"
        >
          Description
        </label>
        <textarea
          id="description"
          placeholder="Insert Description"
          rows="3"
          style="min-height: 150px;"
        >
          Lorem ipsum dolor sip amet
        </textarea>
      </div>
      <p>
        Timer ends in: 
        <span>
          5
        </span>
        <span
          style="margin-left: 8px;"
        />
      </p>
      <div
        class="fields"
      >
        <div
          class="field"
        >
          <button
            class="ui green button"
            type="button"
          >
            Start Timer
          </button>
        </div>
        <div
          class="field"
        >
          <button
            class="ui yellow button"
            type="button"
          >
            Pause Timer
          </button>
        </div>
        <div
          class="field"
        >
          <button
            class="ui red button"
            type="button"
          >
            Stop Timer
          </button>
        </div>
      </div>
      <code
        data-testid="peopleLength"
      >
        0
      </code>
    </form>
  </div>
</body>
`;
